### PR TITLE
Integrate BMS cell balancing state 

### PR DESCRIPTION
**Pull request by pasitlaothamatas**
### Summary

- Add a balancing state dedicated to balancing cell voltages
(1) Balancing allowed when pack is unbalanced and the min voltage is below a given voltage threshold (3.8V)
(2) Once the cell voltage is above that threshold balance until the difference between the max cell voltage and the min cell voltage is below a given threshold (0.02V)? (set to 600uV for now)

- Add be overridden via CAN

### Changelist 

- Using "discharge" and "balancing" when they are supposed to be used

### Testing Done

- Haven't tested yet, will try to test this evening

### Checklist
*Please change `[ ]` to `[x]` when you are ready.*
- [ ] I have read and followed the code conventions detailed in [README.md](../README.md) (*This will save time for both you and the reviewer!*).
- [ ] If this pull request is longer then **500** lines, I have provided *explicit* justification in the summary above explaining why I *cannot* break this up into multiple pull requests (*Small PR's are faster and less painful for everyone involved!*).

### CODE REVIEW

**Comment by pasitlaothamatas** [comment]
should i name this cell balancing state instead

**Comment by pasitlaothamatas** [comment]
finna start a poll up in here

File: boards/BMS/Src/App/App_Accumulator.c [hunk 1]
```diff
@@ -79,8 +95,19 @@ void App_Accumulator_RunOnTick100Hz(
             // Read cell voltages
             accumulator->read_cell_voltages();
 
-            // TODO: Add broadcast BMS CAN messages when fixed pt. CAN PR has
-            // been merged in
+            // Write to configuration registers
+            accumulator->write_to_cfg_reg();
+
+            // If current state machine state is the charge state, disable
```
Comment (kirons96) [inline boards/BMS/Src/App/App_Accumulator.c:101]: weird that comment has the opposite logic than the line below it, and also just restates the code

**Comment by kirons96** [comment]
why do you need a dedicated discharging state? aren't you going to be discharging while in other states?

also, are you guys keeping the specs up to date while you make significant changes like this?

**Comment by pasitlaothamatas** [comment]
> why do you need a dedicated discharging state? aren't you going to be discharging while in other states?
> 
> also, are you guys keeping the specs up to date while you make significant changes like this?

(1) We will not need to discharge all the time as it takes (5-10)? charge cycles for the voltages to drift apart significantly. The plan is to have this discharge state to discharge our cells every once in a while. The plan is as follows but may change based on future discussion:
- Enter charge state 
- Charge until the max cell reaches ~3.7-3.8V
- If min and max are significantly far apart enter discharge state
- Discharge all cells to the min cell voltage
- After cells are balanced, enter charge state 


(2)  Should probably keep specs updated. Can amend specs in a different PR

**Comment by kirons96** [comment]
> We will not need to discharge all the time as it takes (5-10)? charge cycles for the voltages to drift apart significantly. The plan is to have this discharge state to discharge our cells every once in a while.

charging time is super valuable at competition and in the past prevented us from doing other maintenance/debugging while charging. extending the charging time with a discharge cycle could hurt us if that discharge cycle lands in one of the charge cycles at comp. it seems simpler to always balance cells to some threshold when the contactors are open, which will be the vast majority of the time

thoughts @joethrstn @Eddy13013 @Auhv ?

**Comment by Auhv** [comment]
@kirons96  yes.

i'm assuming the charger handles constant voltage and constant current region appropriately (i.e. has access to the min/max cell voltage readings), therefore we can charge until the highest cell is finished and then make the manual decision to balance if we're not happy with our max state of charge. Balancing should be a separate routine/state all together

File: boards/BMS/Src/App/states/App_InitState.c [hunk 2]
```diff
@@ -31,11 +30,17 @@ static void InitStateRunOnTick100Hz(struct StateMachine *const state_machine)
...
+        struct TractiveSystem *   ts          = App_BmsWorld_GetTractiveSystem(world);
+        struct Airs *             airs        = App_BmsWorld_GetAirs(world);
+        struct Accumulator *      accumulator = App_BmsWorld_GetAccumulator(world);
 
-        if (App_Airs_IsAirNegativeClosed(airs) && (App_TractiveSystem_GetVoltage(ts) < TS_DISCHARGED_THRESHOLD_V))
+        // Cell balancing may be overriden by the injection of a CAN message. This will allow the state machine to idle
+        // in the init state
+        if (App_Airs_IsAirNegativeClosed(airs) && (App_TractiveSystem_GetVoltage(ts) < TS_DISCHARGED_THRESHOLD_V) &&
+            !App_Accumulator_GetPackFullyChargedStatus(accumulator) &&
+            !App_CanRx_CAN_DEBUGGING_SIGNALS_GetSignal_DISABLE_CELL_BALANCING(can_rx))
```
Comment (pasitlaothamatas) [inline boards/BMS/Src/App/states/App_InitState.c:43]: Manual override of discharging via CAN

File: boards/BMS/Src/App/states/App_ChargeState.c [hunk 3]
```diff
@@ -44,23 +50,32 @@ static void ChargeStateRunOnTick100Hz(struct StateMachine *const state_machine)
...
         }
 
-        uint8_t    segment = 0U;
-        uint8_t    cell    = 0U;
-        const bool has_reached_max_v =
-            App_Accumulator_GetMaxVoltage(accumulator, &segment, &cell) > MAX_CELL_VOLTAGE_THRESHOLD;
+        uint8_t    segment           = 0U;
+        uint8_t    cell              = 0U;
+        const bool has_reached_max_v = App_Accumulator_GetMaxVoltage(accumulator, &segment, &cell) > MAX_CELL_V_TARGET;
+        App_CanTx_SetPeriodicSignal_IS_CHARGING_COMPLETE(can_tx, has_reached_max_v);
```
Comment (pasitlaothamatas) [inline boards/BMS/Src/App/states/App_ChargeState.c:56]: maybe should move this to when the pack is balanced and we have reached max voltage

File: boards/BMS/Src/App/states/App_BalanceState.c [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+            next_state = App_GetInitState();
+        }
+        else
+        {
+            bool is_balancing_enabled = true;
+
+            // Overriding cell balancing, transition back into the init state
+            if (App_CanRx_CAN_DEBUGGING_SIGNALS_GetSignal_DISABLE_CELL_BALANCING(can_rx))
+            {
+                App_Accumulator_SetCellBalancingStatus(accumulator, false);
```
Comment (pasitlaothamatas) [inline boards/BMS/Src/App/states/App_BalanceState.c:50]:
```suggestion
                is_balancing_enabled = false;
```

File: boards/BMS/Src/App/states/App_ChargeState.c [hunk 5]
```diff
@@ -44,23 +50,32 @@ static void ChargeStateRunOnTick100Hz(struct StateMachine *const state_machine)
...
+        const float min_cell_v       = App_Accumulator_GetMinVoltage(accumulator, &segment, &cell);
+        const bool  is_pack_balanced = (App_Accumulator_GetMaxVoltage(accumulator, &segment, &cell) - min_cell_v) <=
+                                      CONVERT_100UV_TO_VOLTAGE(CELL_VOLTAGE_DISCHARGE_WINDOW_UV);
+
+        if (is_charger_disconnected || has_charger_faulted)
         {
-            App_SharedStateMachine_SetNextState(state_machine, App_GetFaultState());
+            next_state = App_GetFaultState();
         }
+        else if (has_reached_max_v || (!is_pack_balanced && (min_cell_v > MIN_BALANCING_CELL_VOLTAGE)))
```
Comment (pasitlaothamatas) [inline boards/BMS/Src/App/states/App_ChargeState.c:73]: Balance when we are above 3.8V and the pack is unbalanced

File: boards/BMS/Src/App/states/App_BalanceState.c [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+            bool is_balancing_enabled = true;
+
+            // Overriding cell balancing, transition back into the init state
+            if (App_CanRx_CAN_DEBUGGING_SIGNALS_GetSignal_DISABLE_CELL_BALANCING(can_rx))
+            {
+                is_balancing_enabled = false;
+                next_state           = App_GetInitState();
+            }
+
+            // Set a flag signifying that the pack is NOT fully charged. Remain in discharge state until the pack is
```
Comment (pasitlaothamatas) [inline boards/BMS/Src/App/states/App_BalanceState.c:54]:
```suggestion
            // Set a flag signifying that the pack is NOT fully charged. Remain in discharge state until the pack is fully balanced
```

**Comment by pasitlaothamatas** [comment]
I can optimize this a bit more, will let u guys know when its ready 4 review

File: boards/BMS/Src/App/states/App_InitState.c [hunk 7]
```diff
@@ -31,11 +30,17 @@ static void InitStateRunOnTick100Hz(struct StateMachine *const state_machine)
...
+        struct BmsCanRxInterface *can_rx      = App_BmsWorld_GetCanRx(world);
+        struct TractiveSystem *   ts          = App_BmsWorld_GetTractiveSystem(world);
+        struct Airs *             airs        = App_BmsWorld_GetAirs(world);
+        struct Accumulator *      accumulator = App_BmsWorld_GetAccumulator(world);
 
-        if (App_Airs_IsAirNegativeClosed(airs) && (App_TractiveSystem_GetVoltage(ts) < TS_DISCHARGED_THRESHOLD_V))
+        // Cell balancing may be overriden by the injection of a CAN message. This will allow the state machine to idle
+        // in the init state
+        if (App_Airs_IsAirNegativeClosed(airs) && (App_TractiveSystem_GetVoltage(ts) < TS_DISCHARGED_THRESHOLD_V) &&
+            !App_Accumulator_GetPackFullyChargedStatus(accumulator) &&
```
Comment (pasitlaothamatas) [inline boards/BMS/Src/App/states/App_InitState.c:42]: haha, this makes no sense imma fix it

File: boards/BMS/Src/App/App_Accumulator.c [hunk 8]
```diff
@@ -169,10 +198,17 @@ void App_Accumulator_RunOnTick100Hz(struct Accumulator *const accumulator)
 
             UPDATE_PEC15_ERROR_COUNT(accumulator->read_cell_voltages(), accumulator->num_comm_tries)
 
-            // Write to configuration register to configure cell discharging
-            // TODO: re-enable
-            // accumulator->write_cfg_registers();
-            accumulator->disable_discharge();
+            if (accumulator->is_cell_balancing_enabled)
```
Comment (Eddy13013) [inline boards/BMS/Src/App/App_Accumulator.c:201]: I feel like enabling balancing should be done during run on entry in the balancing state. Same for disabling balancing, should be done in run on exit

File: boards/BMS/Src/App/states/App_BalanceState.c [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    struct BmsCanRxInterface *can_rx      = App_BmsWorld_GetCanRx(world);
+    struct BmsCanTxInterface *can_tx      = App_BmsWorld_GetCanTx(world);
+
+    if (App_AllStatesRunOnTick100Hz(state_machine))
+    {
+        const struct State *next_state           = App_GetBalanceState();
+        bool                is_charging_complete = false;
+        bool                is_balancing_enabled = true;
+
+        if (App_Accumulator_IsPackBalanced(accumulator) && App_Accumulator_HasCellReachedMaxVoltageTarget(accumulator))
```
Comment (Eddy13013) [inline boards/BMS/Src/App/states/App_BalanceState.c:34]: you should call Is_Pack_Fully_Charged from App_Accumulator. its the same logic

File: boards/BMS/Src/App/states/App_BalanceState.c [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+            is_balancing_enabled = false;
+            next_state           = App_GetInitState();
+        }
+
+        App_Accumulator_SetIsCellBalancingEnabled(accumulator, is_balancing_enabled);
+
+        // Update is charging complete status on CAN
+        App_CanTx_SetPeriodicSignal_IS_CHARGING_COMPLETE(can_tx, is_charging_complete);
+
+        App_SharedStateMachine_SetNextState(state_machine, next_state);
```
Comment (Eddy13013) [inline boards/BMS/Src/App/states/App_BalanceState.c:52]: Why is this here? shouldn't we just wait in balancing state until the cells are within the specified range?

File: boards/BMS/Src/App/states/App_ChargeState.c [hunk 11]
```diff
@@ -44,23 +52,31 @@ static void ChargeStateRunOnTick100Hz(struct StateMachine *const state_machine)
             has_charger_faulted = App_Charger_HasFaulted(charger);
         }
 
-        uint8_t    segment = 0U;
-        uint8_t    cell    = 0U;
-        const bool has_reached_max_v =
-            App_Accumulator_GetMaxVoltage(accumulator, &segment, &cell) > MAX_CELL_VOLTAGE_THRESHOLD;
+        if (is_charger_disconnected || has_charger_faulted)
```
Comment (Eddy13013) [inline boards/BMS/Src/App/states/App_ChargeState.c:55]: Are we broadcasting these messages over CAN? If not we should be

File: boards/BMS/Src/App/states/App_ChargeState.c [hunk 12]
```diff
@@ -44,23 +52,31 @@ static void ChargeStateRunOnTick100Hz(struct StateMachine *const state_machine)
...
+        {
+            next_state = App_GetFaultState();
+        }
+        else if (!is_pack_balanced && App_Accumulator_IsBalancingAllowed(accumulator))
```
Comment (Eddy13013) [inline boards/BMS/Src/App/states/App_ChargeState.c:59]: This seems redundant. Shouldn't the `App_Accumulator_IsBalancingAllowed(accumulator)` tell us if the pack is balanced? Or at least within a range of unbalanced that is ok

File: boards/BMS/Src/App/states/App_ChargeState.c [hunk 13]
```diff
@@ -44,23 +52,31 @@ static void ChargeStateRunOnTick100Hz(struct StateMachine *const state_machine)
...
+        {
+            // Start balancing when the pack is unbalanced, and that the minimum cell voltage is
+            next_state = App_GetBalanceState();
+        }
+        else if (is_pack_balanced && has_reached_max_v)
```
Comment (Eddy13013) [inline boards/BMS/Src/App/states/App_ChargeState.c:64]: Again, you designed a function in App_Accumulator to check this

File: boards/BMS/Src/App/states/App_FaultState.c [hunk 14]
```diff
@@ -15,26 +15,15 @@ static void FaultStateRunOnEntry(struct StateMachine *const state_machine)
...
 
 static void FaultStateRunOnTick100Hz(struct StateMachine *const state_machine)
 {
-    App_AllStatesRunOnTick100Hz(state_machine);
-
     struct BmsWorld *world = App_SharedStateMachine_GetWorld(state_machine);
     struct Airs *    airs  = App_BmsWorld_GetAirs(world);
 
-    if (!App_Airs_IsAirNegativeClosed(airs))
+    if (App_AllStatesRunOnTick100Hz(state_machine) && !App_Airs_IsAirNegativeClosed(airs))
```
Comment (Eddy13013) [inline boards/BMS/Src/App/states/App_FaultState.c:26]: These changes seem out of scope. We definitely need to rework the fault state but I think it should be removed from this PR

File: scripts/codegen/CAN/App_CanMsgs.dbc [hunk 15]
```diff
@@ -601,8 +602,9 @@ SG_ DUMMY_AIR_SHUTDOWN : 0|1@1+ (1,0) [0|1] "" DCM
 BO_ 510 DIM_MOTOR_SHUTDOWN_ERRORS: 8 DIM
 SG_ DUMMY_MOTOR_SHUTDOWN : 0|1@1+ (1,0) [0|1] "" DCM
 
-BO_ 600 CAN_DEBUGGING_SIGNALS: 1 CAN_DBG
-SG_ ENABLE_CELL_BALANCING : 0|1@1+ (1,0) [0|1] "" BMS
+BO_ 600 CAN_DEBUGGING_SIGNALS: 1 DEBUG
```
Comment (Eddy13013) [inline scripts/codegen/CAN/App_CanMsgs.dbc:605]: rename to CELL_BALANCING?

**Review by Eddy13013 — CHANGES_REQUESTED** [review]
Need to change "discharge" to "balancing" everywhere in the code.

File: boards/BMS/Src/App/states/App_AllStates.c [hunk 16]
```diff
@@ -1,4 +1,6 @@
 #include "states/App_AllStates.h"
+#include "states/App_BalanceState.h"
```
[the remark this answers is not in the document]
Comment (pasitlaothamatas) [reply boards/BMS/Src/App/states/App_AllStates.c:2]: Used for the `App_GetXState()` functions

File: boards/BMS/Src/App/states/App_ChargeState.c (hunk 11 above)
Comment (pasitlaothamatas) [reply boards/BMS/Src/App/states/App_ChargeState.c:55]: yup

File: boards/BMS/Src/App/states/App_FaultState.c (hunk 14 above)
Comment (pasitlaothamatas) [reply boards/BMS/Src/App/states/App_FaultState.c:26]: its a single line so i think its ok to do this

File: scripts/codegen/CAN/App_CanMsgs.dbc (hunk 15 above)
Comment (pasitlaothamatas) [reply scripts/codegen/CAN/App_CanMsgs.dbc:605]: I want this to be a general message containing signals that we can use to take control of the bus

File: boards/BMS/Src/App/states/App_BalanceState.c (hunk 10 above)
Comment (pasitlaothamatas) [reply boards/BMS/Src/App/states/App_BalanceState.c:52]: next state will only change if we run into the if, and else if statements above. notice how next_state is initialized to App_GetBalanceState()

File: boards/BMS/Src/App/App_Accumulator.c [hunk 17]
```diff
@@ -71,11 +75,14 @@ struct Accumulator *App_Accumulator_Create(
     struct Accumulator *accumulator = malloc(sizeof(struct Accumulator));
```
[the remark this answers is not in the document]
Comment (pasitlaothamatas) [reply boards/BMS/Src/App/App_Accumulator.c:75]: Need to send cmd out when the chip is awake. Alternatively we can wake up the chip, then send the enable/disable balancing commands to the chip. However, this would increase the time spent sending the msg on the SPI bus

File: boards/BMS/Src/App/states/App_BalanceState.c (hunk 9 above)
Comment (pasitlaothamatas) [reply boards/BMS/Src/App/states/App_BalanceState.c]: yup forgot to update, should be updated now

File: boards/BMS/Src/App/states/App_ChargeState.c (hunk 12 above)
Comment (pasitlaothamatas) [reply boards/BMS/Src/App/states/App_ChargeState.c]: thought we needed to use `has_reached_max_v` but realized we dont rlly need it. should be improved now

File: boards/BMS/Src/App/states/App_ChargeState.c (hunk 13 above)
Comment (pasitlaothamatas) [reply boards/BMS/Src/App/states/App_ChargeState.c]: > thought we needed to use `has_reached_max_v` but realized we dont rlly need it. should be improved now

^ addressed here

**Comment by pasitlaothamatas** [comment]
Slight modifications after offline discussions:

- (1) If charger is connected charge until we reach the max voltage
- (2) If max voltage is reached but pack is unbalanced, balance cells
- Repeat from (1) until the pack is balanced
- Chill in init when the pack is fully charged (max voltage reached, and balanced)

File: boards/BMS/Src/App/App_Accumulator.c [hunk 18]
```diff
@@ -149,14 +150,42 @@ float App_Accumulator_GetAvgCellTempDegC(const struct Accumulator *const accumul
...
+    uint8_t cell    = 0U;
+    return accumulator->get_max_cell_voltage(&segment, &cell) >= MAX_CELL_V_TARGET;
+}
+
+bool App_Accumulator_IsBalancingAllowed(struct Accumulator *const accumulator)
 {
-    return accumulator->enable_discharge();
+    uint8_t segment = 0U;
+    uint8_t cell    = 0U;
+    return accumulator->get_min_cell_voltage(&segment, &cell) >= MIN_BALANCING_CELL_VOLTAGE;
```
[the remark this answers is not in the document]
Comment (pasitlaothamatas) [reply boards/BMS/Src/App/App_Accumulator.c]: Yeah hopefully this doesn't pop up but we should design for these situations anyway

**Comment by pasitlaothamatas** [comment]
> Slight modifications after offline discussions:
> 
> * (1) If charger is connected charge until we reach the max voltage
> * (2) If max voltage is reached but pack is unbalanced, balance cells (no more min cell voltage threshold where balancing is allowed)
> * Repeat from (1) until the pack is balanced
> * Chill in init when the pack is fully charged (max voltage reached, and balanced)

Discharge curve doesn't look too flat tbh

image

File: boards/BMS/Src/App/App_Accumulator.c [hunk 19]
```diff
@@ -1,9 +1,11 @@
 #include <stdlib.h>
 #include <assert.h>
 #include "App_Accumulator.h"
+#include "App_SharedMacros.h"
 
 // Max number of PEC15 to occur before faulting
 #define MAX_NUM_COMM_TRIES (3U)
+#define MIN_BALANCING_CELL_VOLTAGE (3.8f)
```
Comment (pasitlaothamatas) [inline boards/BMS/Src/App/App_Accumulator.c:8]:
```suggestion
```

File: boards/BMS/Src/App/states/App_InitState.c [hunk 20]
```diff
@@ -31,14 +30,29 @@ static void InitStateRunOnTick100Hz(struct StateMachine *const state_machine)
...
+            (!App_Accumulator_IsPackFullyCharged(accumulator) || !App_Charger_IsConnected(charger)) &&
+            !App_CanRx_CAN_DEBUGGING_SIGNALS_GetSignal_DISABLE_CELL_BALANCING(can_rx))
         {
-            App_SharedStateMachine_SetNextState(state_machine, App_GetPreChargeState());
+            next_state = App_GetPreChargeState();
         }
+        else if (App_CanRx_CAN_DEBUGGING_SIGNALS_GetSignal_ENABLE_BALANCING_FROM_INIT(can_rx))
+        {
+            // Balancing may be enabled directly from the init state
+            next_state = App_GetBalanceState();
```
Comment (Eddy13013) [inline boards/BMS/Src/App/states/App_InitState.c:52]: Note: Cell balancing is only allowed when contactors are closed
